File: android/app/src/main/kotlin/com/lepos/lepos/domain/services/AppInstallationService.kt
package com.lepos.lepos.domain.services

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * Service to manage the installation state of Mini Apps.
 * This is a simulated in-memory implementation for the prototype.
 */
object AppInstallationService {
    private val _installedApps = MutableStateFlow<Set<String>>(emptySet())
    val installedApps: StateFlow<Set<String>> = _installedApps.asStateFlow()

    fun isAppInstalled(appId: String): Boolean {
        return _installedApps.value.contains(appId)
    }

    fun installApp(appId: String) {
        _installedApps.update { it + appId }
    }

    fun uninstallApp(appId: String) {
        _installedApps.update { it - appId }
    }
}

File: ios/iosApp/Shared/Services/AppInstallationService.swift
import Foundation

final class AppInstallationService {
    static let shared = AppInstallationService()

    private let installedAppsKey = "installed_apps"
    private let defaults = UserDefaults.standard

    private init() {}

    func isAppInstalled(appId: String) -> Bool {
        installedApps.contains(appId)
    }

    func installApp(appId: String) {
        var apps = installedApps
        apps.insert(appId)
        defaults.set(Array(apps), forKey: installedAppsKey)
    }

    func uninstallApp(appId: String) {
        var apps = installedApps
        apps.remove(appId)
        defaults.set(Array(apps), forKey: installedAppsKey)
    }

    private var installedApps: Set<String> {
        Set(defaults.stringArray(forKey: installedAppsKey) ?? [])
    }
}
